core/src/main/java/org/springframework/ldap/core/support/AbstractContextSource.java

```diff
@@ -119,10 +119,16 @@
 
 	public static final String SUN_LDAP_POOLING_FLAG = "com.sun.jndi.ldap.connect.pool";
 
-	private static final String JDK_142 = "1.4.2";
-
 	private DirContextAuthenticationStrategy authenticationStrategy = new SimpleDirContextAuthenticationStrategy();
 
+	public AbstractContextSource() {
```

Comment:
I think this should be moved into `afterPropertiesSet`, too, so that it's all in one place.

For example, in `afterPropertiesSet` you would do:

```java
if (this.contextFactory == null) {
    try {
        this.contextFactory = Class.forName(DEFAULT_CONTEXT_FACTORY);
    } catch (ClassNotFoundException e) {
        throw new IllegalArgumentException("Could not find " + DEFAULT_CONTEXT_FACTORY + " and no context factory specified");
    }
}
```

One thing that's nice about this is that the `ClassNotFoundException` is not generated unless no context factory is specified. Since exceptions are expensive to generate, it's nice when they can be avoided.

Reply:
That was my original idea, yes. However, there will be a slight behavioral change compared to 2.3.3.

```
LdapContextSource ctx = new LdapContextSource();
// afterPropertiesSet is not required.
Class<?> contextFactory = ctx.getContextFactory(); // Version 2.3.3 would return com.sun.jndi.ldap.LdapCtxFactory
```
whereas with the change applied
```
LdapContextSource ctx = new LdapContextSource();
ctx.afterPropertiesSet(); // required otherwise the below will return a null value
Class<?> contextFactory = ctx.getContextFactory(); 
```

Please let me know if that is acceptable and I will update the PR.

Reply:
Would it alleviate your concern if the PR also changed to run this logic in a private getter?

For example, you could add:

```java
private Class<?> ensureContextFactoryName() {
    if (this.contextFactory != null) {
        return this.contextFactory.getName();
    }
    try {
        this.contextFactory = Class.forName(DEFAULT_CONTEXT_FACTORY);
    } catch (ClassNotFoundException e) {
        throw new IllegalArgumentException("Could not find " + DEFAULT_CONTEXT_FACTORY + " and no context factory specified");
    }
    return this.contextFactory.getName();
}
```

Then, in `afterPropertiesSet`, you'd do:

```java
ensureContextFactoryName();
```

And in `setupAnonymousEnv`:

```java
env.put(Context.INITIAL_CONTEXT_FACTORY, ensureContextFactoryName());
```

Reply:
Why not just initialize the `DEFAULT_CONTEXT_FACTORY` in the static code block and do not make any other changes?
```java
    private static final Class<?> DEFAULT_CONTEXT_FACTORY;
    static {
        try {
            DEFAULT_CONTEXT_FACTORY = Class.forName("com.sun.jndi.ldap.LdapCtxFactory");
        } catch (ClassNotFoundException e) {
            throw new RuntimeException(e);
        }
    }
```

Reply:
That's problematic since it errors on something that may not be a problem. The class not being present is only a problem if the application does not set a context factory.

Reply:
@jzheaux The `com.sun.jndi.ldap.LdapCtxFactory` class is always present but not always could be accessed. The `ClassNotFoundException` will never be thrown and catching it is just because it's a checked exception. In other words the `DEFAULT_CONTEXT_FACTORY` will always be present and so will be set the `contextFactory` field by default as well.
Following code example works on Java 16 properly:

```java
public class MainClass {
    private static final Class<?> DEFAULT_CONTEXT_FACTORY;
    static {
        try {
            DEFAULT_CONTEXT_FACTORY = Class.forName("com.sun.jndi.ldap.LdapCtxFactory");
        } catch (ClassNotFoundException e) {
            throw new RuntimeException(e);
        }
    }

    public static void main(String[] args) {
        System.out.println(DEFAULT_CONTEXT_FACTORY.getName());
    }
}
```
Value of the `DEFAULT_CONTEXT_FACTORY` is used in the `AbstractContextSource` only for two things:

1. return value of `getContextFactory()`
2. `env.put(Context.INITIAL_CONTEXT_FACTORY, contextFactory.getName());`

With my version of change all of the above will continue to work properly and no behavior change will be introduced.

Reply:
I see. It's interesting that `Class.forName` works when the direct reference does not. Given that, the PR's constructor won't throw the exception in normal JDK 16 circumstances.